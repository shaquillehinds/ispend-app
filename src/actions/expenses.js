import database from "../firebase/database";

//ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return async (dispatch) => {
    const { description = "", note = "", amount = 0, createdAt = 0 } = expenseData;
    const expense = { description, note, amount, createdAt };
    const res = await database.ref("expenses").push(expense);
    return dispatch(addExpense({ id: res.key, ...expense }));
  };
};

//REMOVE_EXPENSE
export const removeExpense = (id) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const startRemoveExpense = ({ id } = {}) => {
  return async (dispatch) => {
    const removed = await database.ref(`expenses/${id}`).remove();
    dispatch(removeExpense(id));
  };
};

//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id: id,
  updates,
});

export const startEditExpense = (id, updates) => {
  return async (dispatch) => {
    await database.ref(`expenses/${id}`).update(updates);
    dispatch(editExpense(id, updates));
  };
};
//REMOVE_ALL_EXPENSES
export const removeAllExpenses = () => ({
  type: "REMOVE_ALL_EXPENSES",
});
