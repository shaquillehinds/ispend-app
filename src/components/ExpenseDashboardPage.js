import React from "react";
const ExpenseDashboardPage = () => (
  <div>
    <div>This is from my dashboard component</div>
  </div>
);

export default ExpenseDashboardPage;
