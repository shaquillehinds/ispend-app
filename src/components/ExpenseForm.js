import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";
import { history } from "../routers/AppRouter";
// import { connect } from "react-redux";

// const date = new Date();

export default class ExpenseForm extends React.Component {
  state = {
    description: "",
    note: "",
    amount: "",
    createdAt: moment(),
    calendarFocused: false,
    error: undefined,
    ...this.props.expenses,
  };

  componentDidMount() {
    this.props.expense &&
      this.setState(() => ({
        description: this.props.expense.description,
        note: this.props.expense.note,
        amount: (this.props.expense.amount / 100).toString(),
        createdAt: moment(this.props.expense.createdAt),
      }));
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  cancelExpense = () => history.push("/dashboard");
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Please provide description and amount" }));
    } else {
      this.setState(() => ({ error: undefined }));
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
      });
    }
  };
  render() {
    return (
      <div className="content-container">
        <form className="form" onSubmit={this.onSubmit}>
          {this.state.error && <p className="form__error">{this.state.error}</p>}
          <input
            className="text-input"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            type="text"
            placeholder="Description"
            autoFocus
          />
          <input
            className="text-input"
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            className="textarea"
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder="Add a note for your expense (Optional)"
          ></textarea>
          <div className="button-container">
            <div>
              <input className="button" type="submit" value="Save Expense" />
            </div>
            <div>
              <button className="button button--dark-grey" onClick={this.cancelExpense}>
                Cancel
              </button>
            </div>
          </div>
        </form>
      </div>
    );
  }
}

// export default connect()(ExpenseForm);
