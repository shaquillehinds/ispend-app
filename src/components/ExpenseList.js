import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
  <div>
    <table>
      <thead>
        <tr>
          <th>Expense</th>
          <th>Amount</th>
          <th>CreatedAt</th>
        </tr>
      </thead>
      <tbody>
        {props.expenses.length === 0 ? (
          <tr>
            <td>No Expenses</td>
          </tr>
        ) : (
          props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense} />)
        )}
      </tbody>
    </table>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
