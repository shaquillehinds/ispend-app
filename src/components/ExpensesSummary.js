import React from "react";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import getTotalExpenses from "../selectors/expenses-total";
import numeral from "numeral";

const ExpensesSummary = (props) => (
  <div>
    <h3>Expenses Total</h3>
    {props.expenses.length > 0 ? (
      <p>
        Viewing {props.expenses.length} {props.expenses.length > 1 ? "expenses" : "expense"} totalling{" "}
        {numeral(getTotalExpenses(props.expenses) / 100).format("$0,0[.]00")}
      </p>
    ) : (
      <p>No Expenses To Add Up</p>
    )}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
