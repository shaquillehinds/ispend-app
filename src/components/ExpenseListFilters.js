import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from "react-dates";
import { v4 as uuid } from "uuid";

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };
  textChangeHandler = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  sortChangeHandler = (e) => {
    e.target.value === "date" ? this.props.sortByDate() : this.props.sortByAmount();
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            {" "}
            <input
              className="text-input"
              type="text"
              placeholder="Search Expenses"
              value={this.props.filters.text}
              onChange={this.textChangeHandler}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.sortChangeHandler}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            {" "}
            <DateRangePicker
              startDate={this.props.filters.startDate}
              startDateId={uuid()}
              endDate={this.props.filters.endDate}
              endDateId={uuid()}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({ filters: state.filters });
const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (value) => dispatch(setTextFilter(value)),
  sortByDate: (value) => dispatch(sortByDate(value)),
  sortByAmount: (value) => dispatch(sortByAmount(value)),
  setStartDate: (value) => dispatch(setStartDate(value)),
  setEndDate: (value) => dispatch(setEndDate(value)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
