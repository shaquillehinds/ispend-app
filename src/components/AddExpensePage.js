import React from "react";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";
import { connect } from "react-redux";

const AddExpensePage = (props) => {
  const onSubmit = (expense) => {
    props.startAddExpense(expense);
    props.history.push("/");
  };
  return (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm onSubmit={onSubmit} />
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});
export { AddExpensePage };
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
