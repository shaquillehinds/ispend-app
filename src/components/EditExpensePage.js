import React from "react";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";
const EditExpensePage = (props) => {
  const removeItemHandler = (e) => {
    e.target.innerText === "Confirm" && props.dispatch(removeExpense({ id: props.match.params.id }));
    e.target.innerText === "Confirm" && props.history.push("/");
    e.target ? (e.target.innerText = "Confirm") : null;
  };
  return (
    <div>
      <h1>Edit Expense</h1>
      <ExpenseForm
        onSubmit={(expense) => {
          const id = props.match.params.id;
          props.dispatch(editExpense(id, expense));
          props.history.push("/");
        }}
        expense={props.expenses}
      />
      <button onClick={removeItemHandler}>Remove</button>

      <button onClick={() => props.history.push("/")}>Cancel</button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  expenses: state.expenses.find((expense) => expense.id === props.match.params.id),
});

export default connect(mapStateToProps)(EditExpensePage);
