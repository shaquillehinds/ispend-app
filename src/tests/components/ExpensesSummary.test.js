import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import { expenses } from "../fixtures/expenses";

test("should render expensesSummary correct with 1 expense", () => {
  const expense = [expenses[1]];
  const wrapper = shallow(<ExpensesSummary expenses={expense} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render expensesSummary correct with 2 expenses", () => {
  const expense = [expenses[1], expenses[2]];
  const wrapper = shallow(<ExpensesSummary expenses={expense} />);
  expect(wrapper).toMatchSnapshot();
});
