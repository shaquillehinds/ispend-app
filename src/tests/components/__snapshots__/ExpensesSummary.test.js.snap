// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expensesSummary correct with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      <p>
        Viewing 
        <span>
          1
        </span>
         
        expense
         totalling
         
        <span>
          $72
        </span>
      </p>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render expensesSummary correct with 2 expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      <p>
        Viewing 
        <span>
          2
        </span>
         
        expenses
         totalling
         
        <span>
          $102
        </span>
      </p>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
